src/burnchains/affirmation.rs

```diff
@@ -0,0 +1,3315 @@
+
+use std::cmp;
+use std::collections::{BTreeMap, HashMap, HashSet, VecDeque};
+use std::convert::{TryFrom, TryInto};
+use std::fmt;
+use std::sync::mpsc::SyncSender;
+use std::time::Duration;
+
+use burnchains::{
+    db::{BurnchainBlockData, BurnchainDB, BurnchainDBTransaction, BurnchainHeaderReader},
+    Address, Burnchain, BurnchainBlockHeader, Error, PoxConstants, Txid,
+};
+use chainstate::burn::{
+    db::sortdb::SortitionDB,
+    operations::leader_block_commit::{RewardSetInfo, BURN_BLOCK_MINED_AT_MODULUS},
+    operations::BlockstackOperationType,
+    operations::LeaderBlockCommitOp,
+    BlockSnapshot, ConsensusHash,
+};
+use util::db::DBConn;
+use util::db::Error as DBError;
+
+use core::StacksEpochId;
+
+use crate::types::chainstate::{
+    BlockHeaderHash, BurnchainHeaderHash, PoxId, SortitionId, StacksAddress, StacksBlockHeader,
+    StacksBlockId,
+};
+use crate::util::boot::boot_code_id;
+
+/// Affirmation map entries.  By building on a PoX-mined block,
+/// a PoB-mined block (in a PoX reward cycle),
+/// or no block in reward cycle _i_, a sortition's miner
+/// affirms something about the status of the ancestral anchor blocks.
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub enum AffirmationMapEntry {
+    PoxAnchorBlockPresent,
+    PoxAnchorBlockAbsent,
+    Nothing,
+}
+
+impl AffirmationMapEntry {
+    pub fn parse(s: &str) -> Option<AffirmationMapEntry> {
+        if s.len() != 1 {
+            return None;
+        }
+
+        for chr in s.chars() {
+            let next = match chr {
+                'p' => AffirmationMapEntry::PoxAnchorBlockPresent,
+                'a' => AffirmationMapEntry::PoxAnchorBlockAbsent,
+                'n' => AffirmationMapEntry::Nothing,
+                _ => {
+                    return None;
+                }
+            };
+            return Some(next);
+        }
+        return None;
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct AffirmationMap {
+    affirmations: Vec<AffirmationMapEntry>,
+}
+
+impl fmt::Display for AffirmationMapEntry {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            AffirmationMapEntry::PoxAnchorBlockPresent => write!(f, "p"),
+            AffirmationMapEntry::PoxAnchorBlockAbsent => write!(f, "a"),
+            AffirmationMapEntry::Nothing => write!(f, "n"),
+        }
+    }
+}
+
+impl fmt::Debug for AffirmationMapEntry {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&format!("{}", &self))
+    }
+}
+
+impl fmt::Display for AffirmationMap {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "AM(")?;
+        for entry in self.affirmations.iter() {
+            write!(f, "{}", &entry)?;
+        }
+        write!(f, ")")
+    }
+}
+
+impl fmt::Debug for AffirmationMap {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&format!("{}", &self))
```

Comment:
I think you can remove a clone with `write!(f, "{}", self)`

Reply:
Done